File: LowLevelDesign/java/task-scheduler/src/com/taskscheduler/ScheduledTask.java
package com.taskscheduler;

// Comparable Interface is built in Java interface that allows object to
// be compared with each other to determine their ordering
// It defines one method int compareTo(T other) which tells java how to compare two objects of type T

import com.taskscheduler.strategy.SchedulingStrategy;
import com.taskscheduler.task.Task;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.UUID;

public class ScheduledTask implements Comparable<ScheduledTask> {

    private LocalDateTime nextExecutionTime;
    private LocalDateTime lastExecutionTime;
    private final String id;
    private final Task task;
    private final SchedulingStrategy strategy;

    public ScheduledTask(Task task, SchedulingStrategy strategy) {
        this.id = UUID.randomUUID().toString();
        this.task = task;
        this.strategy = strategy;
        updateNextExecutionTime();
    }

    public void updateNextExecutionTime() {
        Optional<LocalDateTime> nextTime = strategy.getNextExecutionTime(this.lastExecutionTime);
        this.nextExecutionTime = nextTime.orElse(null);
    }

    public void updateLastExecutionTime() {
        this.lastExecutionTime = nextExecutionTime;
    }

    // It returns Negative -> this object < other
    // It returns Positive -> this object > other
    // Zero -> this.object == other
    @Override
    public int compareTo(ScheduledTask other) {
        return this.nextExecutionTime.compareTo(other.nextExecutionTime);
    }

    public LocalDateTime getNextExecutionTime() {
        return nextExecutionTime;
    }

    public Task getTask() {
        return task;
    }

    public String getId() {
        return this.id;
    }

    public boolean hasMoreExecutions() {
        return nextExecutionTime != null;
    }
}

File: LowLevelDesign/task-scheduler/src/ScheduledTask.ts
import { SchedulingStrategy } from "./strategy/SchedulingStrategy";
import { Task } from "./task/Task";
import { randomUUID } from 'node:crypto';

export class ScheduledTask {
    private taskId: string;
    private task: Task;
    private schedulingStrategy: SchedulingStrategy;
    private nextExecutionTime: Date;
    private lastExecutionTime: Date;

    constructor(task: Task, schedulingStrategy: SchedulingStrategy) {
        this.taskId = randomUUID();
        this.task = task;
        this.schedulingStrategy = schedulingStrategy;
        this.lastExecutionTime = null;
        this.nextExecutionTime = this.schedulingStrategy.getNextExecutionTime(this.lastExecutionTime);
    }

    getScheduledTaskId() {
        return this.taskId;
    }

    getNextExecutionTime() {
        return this.nextExecutionTime;
    }

    getLastExecutionTime() {
        return this.lastExecutionTime;
    }

    updateLastExecutionTime() {
        this.lastExecutionTime = this.nextExecutionTime;
    }

    updateNextExecutionTime() {
        this.nextExecutionTime = this.schedulingStrategy.getNextExecutionTime(this.lastExecutionTime)
    }

    getTask() {
        return this.task;
    }

    hasMoreExecutions() {
        return this.nextExecutionTime != null;
    }

    compareTo(other) {
        if (!other) return -1;

        const thisTime = this.getNextExecutionTime();
        const otherTime = other.getNextExecutionTime();

        if (!thisTime && !otherTime) return 0;
        if (!thisTime) return -1;
        if (!otherTime) return 1;

        return thisTime.getTime() - otherTime.getTime();
    }
}